snapshotter/pkg/dmsetup/dmsetup.go

```diff
@@ -0,0 +1,310 @@
+// Caller needs to suspend and resume device if it is active.
+func CreateSnapshot(poolName string, deviceID int, baseDeviceID int) error {
+	_, err := dmsetup("message", poolName, "0", fmt.Sprintf("create_snap %d %d", deviceID, baseDeviceID))
+	return err
+}
+
+// DeleteDevice sends "delete <deviceID>" message to the given thin-pool
+func DeleteDevice(poolName string, deviceID int) error {
+	_, err := dmsetup("message", poolName, "0", fmt.Sprintf("delete %d", deviceID))
+	return err
+}
+
+// RemoveDevice removes a device (see "dmsetup remove")
+func RemoveDevice(deviceName string, force, retry, deferred bool) error {
+	args := []string{
+		"remove",
+	}
+
+	if force {
+		args = append(args, "--force")
+	}
+
+	if retry {
+		args = append(args, "--retry")
+	}
+
+	if deferred {
+		args = append(args, "--deferred")
+	}
+
+	args = append(args, GetFullDevicePath(deviceName))
+
+	_, err := dmsetup(args...)
+	return err
+}
+
+// Info outputs device information (see "dmsetup info").
+// If device name is empty, all device infos will be returned.
+func Info(deviceName string) ([]*DeviceInfo, error) {
+	output, err := dmsetup(
+		"info",
+		"--columns",
+		"--noheadings",
+		"-o",
+		"name,blkdevname,attr,major,minor,open,segments,events",
+		"--separator",
+		" ",
+		deviceName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		lines   = strings.Split(output, "\n")
+		devices = make([]*DeviceInfo, len(lines))
+	)
+
+	for i, line := range lines {
+		var (
+			attr = ""
+			info = &DeviceInfo{}
+		)
+
+		_, err := fmt.Sscan(line,
+			&info.Name,
+			&info.BlockDeviceName,
+			&attr,
+			&info.Major,
+			&info.Minor,
+			&info.OpenCount,
+			&info.TargetCount,
+			&info.EventNumber)
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse line %q", line)
+		}
+
+		// Parse attributes (see "man dmsetup" for details)
+		info.Suspended = strings.Contains(attr, "s")
+		info.ReadOnly = strings.Contains(attr, "r")
+		info.TableLive = strings.Contains(attr, "L")
+		info.TableInactive = strings.Contains(attr, "I")
+
+		devices[i] = info
+	}
+
+	return devices, nil
+}
+
+// Version returns "dmsetup version" output
+func Version() (string, error) {
+	return dmsetup("version")
+}
+
+// GetFullDevicePath returns full path for the given device name (like "/dev/mapper/name")
+func GetFullDevicePath(deviceName string) string {
+	if strings.HasPrefix(deviceName, DevMapperDir) {
+		return deviceName
+	}
+
+	return DevMapperDir + deviceName
+}
+
+// BlockDeviceSize returns size of block device in bytes
+func BlockDeviceSize(devicePath string) (uint64, error) {
+	data, err := exec.Command("blockdev", "--getsize64", "-q", devicePath).CombinedOutput()
+	output := string(data)
+	if err != nil {
+		return 0, errors.Wrapf(err, output)
+	}
+
+	output = strings.TrimSuffix(output, "\n")
+	return strconv.ParseUint(output, 10, 64)
+}
+
+func dmsetup(args ...string) (string, error) {
+	data, err := exec.Command("dmsetup", args...).CombinedOutput()
```

Comment:
Do we really want both `stdout` and `stderr`? Looks like `Table` is the only thing that uses the output, so I am curious what would be the behavior if the output was not of the expected format, like some `stderr` stuff in there?

Reply:
In case of error `dmsetup` and/or `blockdev` will return non zero error code, so `exec.Command` will return an error. `CombinedOutput` is needed because I want to capture all error output and return it in error message, so we have error details

Reply:
Is the output used for logging? Or how is the string to be used?

Reply:
Output is used to add error details, see [here](https://github.com/firecracker-microvm/firecracker-containerd/pull/51/files#diff-dc540783dc366cf76abc1b48546be56eR279) and eventually will be logged by snapshotter or containerd

Reply:
Do we need to return it then? I am just wondering since most calls to `dmsetup` ignore the the first returned parameter.